pilot/pkg/xds/ads.go

```diff
@@ -608,78 +545,69 @@ func (s *DiscoveryServer) pushConnection(con *Connection, pushEv *Event) error {
 
 	// This depends on SidecarScope updates, so it should be called after SetSidecarScope.
 	if !ProxyNeedsPush(con.proxy, pushEv) {
-		if con.proxy.XdsResourceGenerator != nil {
-			// to verify if logic works on generator
-			adsLog.Infof("Skipping generator push to %v, no updates required", con.ConID)
-		} else {
-			adsLog.Debugf("Skipping push to %v, no updates required", con.ConID)
-		}
+		adsLog.Debugf("Skipping push to %v, no updates required", con.ConID)
 
-		if s.StatusReporter != nil {
-			// this version of the config will never be distributed to this envoy because it is not a relevant diff.
-			// inform distribution status reporter that this connection has been updated, because it effectively has
-			for _, distributionType := range AllEventTypes {
-				s.StatusReporter.RegisterEvent(con.ConID, distributionType, pushRequest.Push.Version)
-			}
-		}
+		reportAllEvents(s.StatusReporter, con.ConID, pushRequest.Push.Version, nil)
 		return nil
 	}
 
 	adsLog.Infof("Pushing %v", con.ConID)
 
-	// check version, suppress if changed.
-	currentVersion := versionInfo()
-
-	// When using Generator, the generic WatchedResource is used instead of the individual
-	// 'LDSWatch', etc.
-	// Each Generator is responsible for determining if the push event requires a push -
-	// returning nil if the push is not needed.
-	if con.proxy.XdsResourceGenerator != nil {
-		for _, w := range con.proxy.WatchedResources {
-			err := s.pushGeneratorV2(con, pushRequest.Push, currentVersion, w, pushRequest.ConfigsUpdated)
-			if err != nil {
-				return err
-			}
-		}
-	}
-
-	pushTypes := PushTypeFor(con.proxy, pushEv)
-
-	if con.Watching(v3.ClusterType) && pushTypes[CDS] {
-		err := s.pushCds(con, pushRequest.Push, currentVersion)
+	// Send pushes to all generators
+	//E ach Generator is responsible for determining if the push event requires a push
+	for _, w := range getPushResources(con.proxy.WatchedResources) {
```

Comment:
Can we maintain this order some where may be in Proxy so that we do not do it for every push?  i.e. first iterate the order and if it is there in the WatchedResource - call generate and at the end call the remaining ones?